README.md
APILOT
------
* PREVIEW

  ![image](https://user-images.githubusercontent.com/43668841/234758895-29ce21fa-521a-444f-8318-e00ce991a03c.png)

  ![image](https://user-images.githubusercontent.com/43668841/234758774-851624ee-5f15-4492-bbaf-9239bf5ddb94.png)

  ![image](https://user-images.githubusercontent.com/43668841/231068110-1bd61f5b-8c1f-406b-bacd-419e47289a08.png)
  
  ![image](https://user-images.githubusercontent.com/43668841/231068378-ebbcaafe-55d1-4e18-a08c-8dd01a337a7f.png)
  
  ![image](https://user-images.githubusercontent.com/43668841/231068460-11c7977a-feaa-4f1c-b4be-263d4132d213.png)

* 읽어보기
  * 설정값을 함부로 건들면 사고의 위험이 있으며, 사고/고장 발생시 본인의 책임입니다.
  * Panda코드를 수정하였으므로, comma connect에 연결되면 BAN당함(연결되어도 업로드안됨)!
  * 배선개조: SCC모듈(레이더)의 CCAN연결을 절단 -> 판다의 CAN2에 연결
  * NDA지원 (Neokii)
  * 설치브랜치
    * 콤마3: c3-master
    * 콤마2, 이온/원플러스: c2-master
  * 신호정지모드
    * 신호정지감속정지상태에서 GAS페달: 신호무시
    * 소프트 브레이크홀드기능: 정지상태에서 Brake페달 0.7초이상: 신호무시하고 정지유지 
  * 테스트된차량: SCC모듈(레이더) 배선개조 CAN BUS2로 연결된차량
    * Hyundai SANTAFE HYBRID Hybrid 2022 (배선개조 안된것도 가능,단,AEB off됨) (RadarTracks)
    * Kia STINGER
    * Hyundai GENESIS (White PANDA를 이용하여 MDPS개조된차량)
    * KONA EV
    * SANTAFE TM (RadarTracks)
    * 펠리세이드 (RadarTracks)
  * 인게이지
    * 롱컨모드
      * 배선개조된경우: 크루즈셋버튼, (+)(-) 버튼
      * 순정배선(SCC모듈기능정지,AEB OFF): (+)(-) 버튼
      * 자동인게이지 기능
      * 기어선택유무에 관계없이, 안전벨트만 착용하면 인게이지 가능
    * SCC모드
      * 크루즈셋버튼: 차량의 크루즈셋 조건에 따름.
  * 크루즈 차량간격: 속도에 따른 자동차간거리
    * 1단계: 연비절감모드,
    * 2단계: 안전주행모드,
    * 3단계: 일반주행,
    * 4단계: 고속모드 (신호감지안함)
  * NOO Helper (Navigate On Openpilot)
    * 이 기능은 C3의 경우 자체네비, C2의 경우 외부네비(Mappy)를 이용한 조향 및 턴제어 보조 기능입니다(시험용)
    * NOO Helper를 1로 변경하면 작동
    * APilotMan 필수설치(첨부)
    * Mappy 설치 (첨부)
    * PC에 폰을 연결하고 다음을 실행해야함. (필수)
      * adb shell pm grant com.ajouatom.apilotman android.permission.READ_LOGS
        * https://kibua20.tistory.com/165
        * https://forbes.tistory.com/1256
    * 확인방법: Mappy실행 후 목적지설정하고 나서, ApilotMan화면에 메시지출력이 나오면 정상작동

* 설정: 토글
  * Experimental openpilot Longitudinal Control(롱컨트롤)
    * 순정 SCC의 기능을 사용하지 않고 오픈파일럿의 크루즈제어 기능 선택
    * 배선 개조시 선택함.
    * 배선 비개조시, Enable RadarTracks(ON)한경우: 테스트차량만 가능
      * 지원: SANTAFE HEV, NEXO
      * 미지원: GENESIS(DH), KONA
  * SCC Module connected BUS2
    * 배선을 개조하여 판다의 BUS2에 연결한 경우 ON
  * EnableRadarTracks
    * SCC모듈에서 제공하는 레이더정보를 무시하고, SCC모듈내의 레이더 정보를 이용함.

* 설정: 기타
  * 차량선택: 현대/기아 차량만 선택가능
  * 업데이트: 현재선택되어 있는 브랜치에 대해 새로운 업데이트가 있으면 적용 후 Rebooting
    * (주의) 반드시 시동을 끄고 할것!
  * 과속카메라작동방법
    * 0: 사용안함
    * 1: NDA 사용, Neokii의 NDA지원 단말기 설치시 자동으로 연결: 같은 네트워크에 있어야함.
    * 과속카메라 속도제한은 무조건 작동함.
  * 크루즈소리 
    * 0: 크루즈 작동소리가 안남
    * 1: 일부소리가 남.
    * 2: 모든 크루즈 작동소리가 안남.
  * CustomMapBox입력
    * http://IP주소:8082 접속하여 mapboxtoken을 입력하면 자동으로 켜짐.
    * OFF하면 입력된 Token값을 제거함.
  * Show Debug UI
    * 화면에 디버그 정보를 표시함.
  * 시간정보표시
    * 화면에 시간을 표시합니다.
  * 인게이지 유지모드
    * 롱컨선택시 유효하며, Drive 기어, 안전벨트착용하면 최초 1회에 한하여 자동인게이지 됨.
  * 레인모드속도설정
    * 차선이 보일때 자동으로 차선의 중앙으로 따라감.
    * 해당속도이상이 되면 자동으로 자동레인모드로 변경됨
  * 차선치우침 좌우보정
  * 핸들햅팅기능
    * 핸들진동기능이 있는 차량은 과속카메라 감속시 진동을 2번 울려줌.

* 설정: 크루즈
  * 가속시 크루즈속도를 맞춤
    * 가속페달을 밟아 설정된 속도보다 높아지면 자동으로 설정속도를 올려줌
  * 자동속도증가모드(100%)
    * 선행차량의 속도가 빨라지면, RoadSpeedLimit * 비율의 속도까지 자동으로 설정속도를 올려줌.
    * 0: 사용안함
  * 선행차속도에 크루즈속도맞추기(+40km/h)
    * 선행차를 만나면 선행차의 속도에 설정속도로 미리 낮추어줌.
  * 운전모드 초기값(3)
    * 1: 연비모드 : 가속을 제한하여 연비위주의 운전모드
    * 2: 안전모드 : 일반주행보다 더 천천히 가속, 천천히 감속, 차량을 멀리 떨어져서 주행하도록 함.
    * 3: 일반주행모드 : 일반적인 주행모드
    * 4: 고속모드 : 신호정지 기능을 사용안함.
    * 화면의 거리조정(GAP) 표시를 누르면 바뀜.
    * 차간거리조정(GAP)스위치를 길게 누르면 바뀜
  * 드라이브모드: 연비가속비율
    * 연비모드 선택시 일반모드가속대비 비율
  * 드라이브모드: 안전비율
    * 가속도, 차간거리, 감속율등을 안전하게 비율대로 줄이거나 늘려줌.
  * 크루즈버튼작동모드 (Cruise Button Mode)
    * 0: 일반속도제어  (Normal Mode)
      * (+)(-) 1씩증감, 길게누르면 10씩증감 (short: 1,long: +-10)
      * (-)버튼: 현재속도로 설정 (current speed set)
    * 1: 사용자속도제어1 (User1)
      * (+)버튼: 도로속도제한까지 한번에 올라감, 이후 +10km/h씩올라감. (upto road speed limit, after +10)
      * (-)버튼: 현재속도로 설정 (current speed set)
    * 2: 사용자속도제어2 (User2)
      * 사용자속도제어1의 기능외에 (include User1)
      * (-)버튼: 또 누르면, 관성제어모드(크루즈꺼짐)  ( twrice (-) button: cruise off, but leadcar or traffic signal detected cruise on automatically)
    * 3: 사용자속도제어3 (User3)
      * (+)버튼: 1씩증가, 길게 10씩증가  (short:1, long: +10)
      * (-)버튼: 현재속도셋, 10씩 감소, 길게 10씩감소 (-: current speed, after -10)
    * 4: 사용자속도제어4 (User4)
      * (+)버튼: 1씩증가, 길게 10씩증가 (shoft:+1, long: +10)
      * (-)버튼: 10씩 감소, 길게 10씩감소 (shoft: -10, long: -10))
  * 엑셀크루즈ON: 60%이상 엑셀을 밟으면 크루즈가 켜짐.
  * 엑셀크루즈ON: 속도(30): 지정된 속도가 올라가면 크루즈가 켜짐
  * 엑셀크루즈ON: 속도설정방법
    * 0: 현재속도로 세팅
    * 1: 기존 속도로 세팅
    * 2: 선행차 있을때만 기존속도로 세팅
  * 브레이크해제 크루즈ON  사용: 브레이크를 떼면 크루즈를 켜는 기능을 사용
  * 브레이크해제 크루즈ON:주행중,선행차(20): 선행차거리: 브레이크를 떼고 선행차가 일정거리 이상이면 크루즈 ON
  * 브레이크해제 크루즈ON:정지상태, 선행차: 정지상태에서 선행차가 10M이내에 있으면 크루즈 ON
  * 브레이크해제 크루즈ON:주행중, 속도: 0:사용안함, 브레이크를 떼고 일정속도이상이면 크루즈를 현재속도로 ON
  * 브레이크해제 크루즈ON:주행중,신호: 60km/h이하에서 브레이크를 떼고 신호가 감지되면 현재속도로 크루즈 ON

* 설정: 튜닝
  * StopDistance(600cm): 선행차에 대한 정지위치를 조정
  * 신호정지 위치 조정(200cm): 신호등에 대한 정지위치 조정
  * 신호정지 모델속도(OFF): 신호등에 대한 정지시 모델이 제공하는 속도에 따름
    * 사용안함.
  * 롱컨: JERK값(8): 값을 크게하면 가속/감속에 대한 차량의 반응이 강해지나, 브레이크작동이 잘 안되면 값을 조금씩 올림.
  * 롱컨: FF게인(106%): 속도플래너의 가속도에 대한 피드포워드, 이값은 조정하시 말것!
  * 롱컨: P게인(100): 차량 속도제어에 대한 P게인, 50~100정도로 조정.
  * X_EGO_COST(6): 높으면 신호정지위치에 정밀하게 서려고함.
  * J_EGO_COST(5): 고정
  * A_CHANGE_COST(150): 끼워들기 차량에 대해 반응이 너무 강하면 올릴것. (50~200)
  * DANGER_ZONE_COST(100): 고정
  * 차량간격유지 동적제어(OFF): 전방차량이 멀어지는 것에 대한 반응을 빠르게 COST를 동적으로 적용함.
  * 차량간격 동적제어:상대속도(110%): 선행차와의 상대속도에 따라 차량간격을 동적으로 조정함.
  * 차량간격 동적제어:감속(110%): 차량의 감속도에 따라 차량간격을 동적으로 조정함. 급감속을 하면 더 멀리.
  * 차량간격 비율(100%)
    * 위험: 추돌의 위험이 있음.
    * 속도에 따라 차량간격을 자동조정하고 있으나, 좀더 가깝게 하고 싶을때
  * 신호감지 감속율(80%): 신호등감지시 감속도를 조절하여 서서히 속도를 줄이고 싶을때
  * 모델의 자동속도조절의 적용속도(0): 사용안함.
  * 초기가속제한속도(3km/h): 설정속도까 연비운전모드의 가속률을 사용함.
  * 모델혼잡시 조향가속비율적용(ON): 일반적으로 도로상황이 안좋을때, 연비운전모드의 가속률을 사용함.
  * 가속도 제어(100): 가속도를 좀더 올리고 싶을때 사용.
  * 모델커브속도조절(ON): 곡선도로를 만나면 자동으로 속도를 줄여줌.

* 엑셀에서 발을 떼었을때
  * 크루즈ON 조건 
    * 엑셀크루즈ON 설정상태
    * 핸들20도이내이고 엑셀크루즈ON속도이상 또는 가속페달을 60%이상 밟았을때
    * 적색 신호감지는 안되어있는 상태 
    * 속도세트
      * 60%이상밟았을때 기존속도
      * 엑셀크루즈ON: 속도복원설정에 따라 속도복원
  * 크루즈OFF조건
    * 소프트홀드
    * 신호감지 감속중
    * 저속주행
    * 설정속도 보다 느리고, 앞에 차가 있고, 턴중
  * 크루즈 속도변경
    * 페달을 0.6초이내에 뗀경우
    * 크루즈속도보다 높을때

* 브레이크 크루즈ON/OFF/속도조절
  * 크루즈ON조건
    * 소프트홀드
    * 직진, 선행차가 일정거리 이상, 깜박이 꺼짐
    * 직진, 속도 70키로이내, 적색신호등감지, 깜박이 꺼짐
    * 선행차없음, 일정속도이상인경우
    * 정지중, 선행차 10M(설정값)이내인경우

* 설정값참조: SANTAFE_TM_HEV (SCC배선개조)
  * 조향
    * 레인모드작동속도: 0
    * 자동모델턴제어: 1
    * 자동모델턴제어: 속도 : 45
    * 자동모델턴제어:시간제한 : 240
    * 자동차선변경속도: 30
    * 차선치우침좌우보정: 0
  * 정지
    * 가속초기속도 : 0
    * 신호정지 위치조정: 200
    * 정지거리: 550
    * 정지유지브레이크: 30
  * 조향튜닝
    * LiveSteerRatioApply: 100
    * SteerActuatorDelay:30
    * SteerDeltaDown: 7
    * SteerDeltaUp: 4
    * SteeringRateCost: 10
    * _LateralAccelCost: 100
    * _LateralJerkCost: 5
    * _LateralMotionCost: 9
    * _PathCostApply: 100
  * 주행설정
    * 안전비율: 80
    * 연비가속비율: 80
  * 가속설정
    * 가속설정1: 200
    * 가속설정2: 140
    * 가속설정3: 80
    * 가속설정4: 70
    * 가속설정5: 60
    * 가속설정6: 30
  * 주행튜닝
    * A_CHANGE_COST: 200
    * DANGER_ZONE_COST: 100
    * FF게인: 100
    * JERK값: 8
    * KiV게인: 500
    * KpV게인: 100
    * LEAD_DANGER_FACTOR: 85
    * _LADLowerBound: 10
    * _LADUpperBound: 70
    * 신호정지 감속모델선택: 30
    * 신호정지 감속율: 80
  * 감속제어
    * 과속카메라작동방법: 1
    * 과속카메라감속시작시간: 15
    * 과속카메라감속완료시간: 8
    * 커브속도: 1
    * 커브속도제어INDEX: 15
    * 커브속도조절비율: 130
  * 버튼설정
    * 크루즈갭: 초기값: 4
    * 크루즈갭버튼 작동모드: 2
    * 크루즈버튼작동모드: 2
  * 자동크루즈
    * 브레이크해제 크루즈ON:사용: 1
    * 브레이크해제 크루즈ON:정지: 1
    * 브레이크해제 크루즈ON:주행,선행차: 10
    * 브레이크해제 크루즈ON:주행,속도: 40
    * 브레이크해제 크루즈ON,주행,신호: 1
    * 엑셀크루즈OFF:모드: 1
    * 엑셀크루즈ON: 사용: 1
    * 엑셀크루즈ON: 복원속도: 0
    * 엑셀크루즈ON: 속도: 35
    * 자동속도업데이트속도: 130
    * 자동속도증가모드: 130
    * 크루즈최저속도: 15
  * 화면
    * 화면TPMS: 1
    * 화면가속도표시: 0
    * 화면디바이스정보: 0
    * 화면레이더감지정보: 2
    * 화면부가정보: 0
    * 화면차선정보: 0
    * 화면패쓰끝표시: 0
    * 화면표시모드: 2
    * 화면핸들표시모드: 2
    * 화면핸들회전: 1
  * 시작
    * SCC배선개조 BUS2연결: 1
    * 드라이브모드: 초기값: 3
    * 레이더트랙사용: 1
    * 롱컨사용: 1
    * 소프트오토홀드기능: 2
    * 인게이지유지모드: 1
    * 자동인게이지사용: 1
    * 핸들햅팅기능사용: 0
  * 차량간격
    * 감속: 105
    * 차량간격: 92
    * 상대속도: 102
    * 차량간격비율: 95
    * 차량간격유지 동적제어사용: 1
  

![](https://i.imgur.com/b0ZyIx5.jpg)

Table of Contents
=======================

* [What is openpilot?](#what-is-openpilot)

RELEASES.md
230711 - APILOT
========================
* NOO Helper - 0:None,1:Openpilot,2:Mappy
* HDA Speed limit
* 일부 차량 가감속 불량해결(KONA_EV외)


230528 - APILOT
========================
* 크루즈버튼 속도단위설정
* UI일부수정

230525-1 - APILOT
========================
* Torque AccelFactor, Friction수동설정 (조향튜닝)
* StartAccel(가속초기속도) 적용시 : Jerk: 1
* UI일부수정
* 신호감지기능수정

230525 - APILOT
========================
* 추월가속시 전방차량발견 거리 수정
* 커브속도조절비율진입: 시험중..
* 3분마다 liveTorqueData 저장
* 신호인식 mix

230522 - APILOT
========================
* deviceState : 표시방식 변경
* 추월가속시 전방차량발견하면 롱컨OFF
* AlertEvent 표시방식 변경

221216 - APILOT
========================
* 안전주행모드(겨울철)
* 각 모드별 가속기능 정리
* 차간거리 1,2,3 오토단계
* 순정 브레이크홀드 사용
* 기타 UI수정

221210 - APILOT
========================
* 크루즈 버튼모드3,4
* 오토홀드기능 혼합사용
* 순정네비 차량감속기능

221203 - APILOT
========================
* startAccel, stopAccel 설정
* 정지선 필터링

221201 - APILOT
========================
* 차량각격제어를 3가지로 나눔.
  * 차량간격제어: 상대속도+
  * 차량간격제어: 상대속도-
  * 차량간격제어: 감속


221120 - APILOT
========================
* ExperimentalMode: 핸들(GPS) 아이콘을 누르면 토글, 다음 쿠루즈 인게이지시 작동시작함.

221119 - APILOT
========================
* GAP버튼: 선행차와의 거리조정으로 수정
* 신호정지선 filtering: 갑자기 바뀌면 리셋.. 필터리 1.0초
* Dynamic T follow지원 : 범위수정
* 소프트홀드상태에서 신호바뀌면: 전방주의 알림 띄우기기능.
* 기타 버그수정.

221118 - APILOT
========================
* 신호정지선 filtering
* Dynamic T follow지원 : 선행차와의속도, 감속도에 따라 선행차와의 거리 자동조정
* 신호감지조건 일부수정
* 신호감지시 모델속도 추종기능 삭제
* 기타 버그수정.

221115 - APILOT
========================
* EnableRadarTracks: SANTAFE_HEV, NEXO 지원
* EnableRadarTracks: 차량등록만하면 지원하도록 수정함.
* WifiAddress표시
* 위험: 선행차량 간격 조절 설정 (차량간격 동적제어기능과 같이 사용할것!)
* Model: vq_driving

221114 - APILOT
========================
* 차량간격(t_follow)동적제어 기능
* Albert Einstein Model 적용
* 롱컨배선개조 자동인식기능 삭제: 설정해야함.
* Vision Curve속도조절기능 수정
* HUD및 계기판 선행차량 표시기능 수정
* 가속페달 크루즈ON조건 0인경우: 악셀을 40%이상 밟으면 이전 크루즈속도 복원, 그외는 현재속도.
* 크루즈버튼모드: 1,2인경우 악셀을 잠깐씩밟으면 속도가 10씩증가(최초 스피드리밋까지 한번에증가), 30KM/H이상 주행시

221107 - APILOT
========================
* 초기가속제한속도 설정(20km/h)

221107 - APILOT
========================
* pcmCruise 순정SCC모듈을 사용한 크루즈 기능 사용가능
* 크루즈셋을 이용한 인게이지 방식 (크루즈셋버튼: SCC모듈을 셋하고 인게이지, +-버튼: SCC모듈을 사용안하고 인게이지): 별로 차이는 없어보임.

221105 - APILOT
========================
* EnableRadarTracks메뉴: SCC11의 값을 이용하지 않고, 직접 레이더 정보를 수신(현재, SANTAFE 2022 HEV만 지원)
* 전방주의 계기판
* 신호감지시 감속율조정
* SCC순정기능 (ExperimentalLonitudinal 방식사용안함): 단, 기어변속시 에러발생
* 140키로이상 속도가 안올라는 현상 수정
* LongDynamicCost

dev221102 - APILOT
========================
* 신호감속정지시 모델의 속도를 참조하도록 함(설정)
* 속도크루즈ON:속도설정방법: 엑셀을 밟아 자동인게이지가 될때, 세트속도설정방법
* 엑셀크루즈ON: 코드반영이 안되었음: 60%이상의 엑셀을 밟으면 속도와 관계없이 크루즈ON

Version 0.8.17 (2022-XX-XX) - APILOT
========================
* New driving model
  * Internal feature space accuracy increased tenfold during training, this makes the model dramatically more accurate.
* Self-tuning torque lateral controller parameters
  * Parameters are learned live for each car

Version 0.8.16 (2022-08-26)
========================
* New driving model
  * Reduced turn cutting
* Auto-detect right hand drive setting with driver monitoring model
* Improved fan controller for comma three
* New translations
  * Japanese thanks to cydia2020!
  * Brazilian Portuguese thanks to AlexandreSato!
* Chevrolet Bolt EUV 2022-23 support thanks to JasonJShuler!
* Chevrolet Silverado 1500 2020-21 support thanks to JasonJShuler!
* GMC Sierra 1500 2020-21 support thanks to JasonJShuler!
* Hyundai Ioniq 5 2022 support thanks to sunnyhaibin!
* Hyundai Kona Electric 2022 support thanks to sunnyhaibin!
* Hyundai Tucson Hybrid 2022 support thanks to sunnyhaibin!
* Subaru Legacy 2020-22 support thanks to martinl!
* Subaru Outback 2020-22 support

Version 0.8.15 (2022-07-20)
========================
* New driving model
  * Path planning uses end-to-end output instead of lane lines at all times
  * Reduced ping pong
  * Improved lane centering
* New lateral controller based on physical wheel torque model
  * Much smoother control that's consistent across the speed range
  * Effective feedforward that uses road roll
  * Simplified tuning, all car-specific parameters can be derived from data
  * Used on select Toyota and Hyundai models at first
  * Significantly improved control on TSS-P Prius

selfdrive/manager/manager.py
#!/usr/bin/env python3
import datetime
import os
import signal
import subprocess
import sys
import traceback
from multiprocessing import Process
from typing import List, Tuple, Union

import cereal.messaging as messaging
import selfdrive.sentry as sentry
from common.basedir import BASEDIR
from common.params import Params, ParamKeyType
from common.text_window import TextWindow
from selfdrive.boardd.set_time import set_time
from selfdrive.hardware import HARDWARE, PC, EON
from selfdrive.manager.helpers import unblock_stdout
from selfdrive.manager.process import ensure_running, launcher
from selfdrive.manager.process_config import managed_processes
from selfdrive.athena.registration import register, UNREGISTERED_DONGLE_ID
from selfdrive.swaglog import cloudlog, add_file_handler
from selfdrive.version import is_dirty, get_commit, get_version, get_origin, get_short_branch, \
                              terms_version, training_version
from selfdrive.hardware.eon.apk import system

sys.path.append(os.path.join(BASEDIR, "pyextra"))


def manager_init() -> None:
  # update system time from panda
  set_time(cloudlog)

  # save boot log
  #subprocess.call("./bootlog", cwd=os.path.join(BASEDIR, "selfdrive/loggerd"))

  params = Params()
  params.clear_all(ParamKeyType.CLEAR_ON_MANAGER_START)

  default_params: List[Tuple[str, Union[str, bytes]]] = [
    ("CompletedTrainingVersion", "0"),
    ("DisengageOnAccelerator", "0"),
    ("GsmMetered", "1"),
    ("HasAcceptedTerms", "0"),
    ("LanguageSetting", "main_en"),
    ("OpenpilotEnabledToggle", "1"),
    ("ShowDebugUI", "0"),
    ("OPKRServer", "0"),
    ("ShowDateTime", "1"),
    ("ShowHudMode", "4"),
    ("ShowSteerRotate", "1"),
    ("ShowPathEnd", "1"),
    ("ShowAccelRpm", "1"),
    ("ShowTpms", "1"),
    ("ShowSteerMode", "2"),
    ("ShowDeviceState", "1"),
    ("ShowConnInfo", "1"),
    ("ShowLaneInfo", "1"),
    ("ShowBlindSpot", "1"),
    ("ShowGapInfo", "1"),
    ("ShowDmInfo", "1"),
    ("ShowRadarInfo", "1"),
    ("MixRadarInfo", "0"),
    ("ShowZOffset", "122"),
    ("ShowPathMode", "9"),
    ("ShowPathColor", "12"),
    ("ShowPathModeCruiseOff", "0"),
    ("ShowPathColorCruiseOff", "19"),
    ("ShowPathModeLane", "12"),
    ("ShowPathColorLane", "13"),
    ("ShowPathWidth", "100"),
    ("ShowPlotMode", "0"),
    ("AutoResumeFromGas", "1"),
    ("AutoResumeFromGasSpeed", "30"),
    ("AutoResumeFromGasSpeedMode", "0"),    
    ("AutoCancelFromGasMode", "1"),    
    ("OpkrPrebuiltOn", "0"),
    ("OpkrHotspotOnBoot", "0"),
    ("OPKRTimeZone", "Asia/Seoul"),    
    ("AutoCurveSpeedCtrlUse", "1"),
    ("AutoCurveSpeedFactor", "100"),
    ("AutoCurveSpeedFactorIn", "10"),
    ("AutoTurnControl", "0"),
    ("AutoTurnSpeed", "40"),
    ("AutoTurnTimeMax", "200"),
    ("AutoLaneChangeSpeed", "30"),
    ("AutoNaviSpeedCtrl", "1"),
    ("AutoNaviSpeedCtrlStart", "22"),
    ("AutoNaviSpeedCtrlEnd", "6"),
    ("AutoRoadLimitCtrl", "0"),
    ("AutoResumeFromBrakeRelease", "1"),
    ("AutoResumeFromBrakeReleaseDist", "20"),
    ("AutoResumeFromBrakeReleaseLeadCar", "1"),
    ("AutoResumeFromBrakeCarSpeed", "40"),
    ("AutoResumeFromBrakeReleaseTrafficSign", "1"),
    ("XEgoObstacleCost", "6"),
    ("JEgoCost", "5"),
    ("AChangeCost", "200"),
    ("DangerZoneCost", "100"),
    ("LeadDangerFactor", "80"),
    ("LongControlActiveSound", "1"),
    ("StartAccelApply", "0"),
    ("StopAccelApply", "30"),
    ("TrafficStopDistanceAdjust", "400"),
    ("AutoSpeedUptoRoadSpeedLimit", "100"),
    ("ApplyLongDynamicCost", "0"), 
    ("AutoSpeedAdjustWithLeadCar", "0"),   
    ("TrafficStopAccel", "80"),     
    ("TrafficStopModelSpeed", "0"),         
    ("TrafficStopMode", "1"),         
    ("CruiseButtonMode", "0"),      
    ("CruiseSpeedUnit", "10"),      
    ("GapButtonMode", "0"),      
    ("InitMyDrivingMode", "3"),      
    ("MySafeModeFactor", "80"),      
    ("LiveSteerRatioApply", "100"),      
    ("MyEcoModeFactor", "80"),  
    ("CruiseMaxVals1", "160"),
    ("CruiseMaxVals2", "120"),
    ("CruiseMaxVals3", "100"),
    ("CruiseMaxVals4", "80"),
    ("CruiseMaxVals5", "70"),
    ("CruiseMaxVals6", "60"),
    ("PrevCruiseGap", "4"),      
    ("CruiseSpeedMin", "10"),
    ("AutoSyncCruiseSpeedMax", "120"),       
    ("StopDistance", "600"), 
    ("CustomMapbox", "0"),    
    ("LongitudinalTuningKpV", "100"),     
    ("LongitudinalTuningKiV", "200"),     
    ("LongitudinalTuningKf", "100"),     
    ("LongitudinalActuatorDelayUpperBound", "50"),     
    ("LongitudinalActuatorDelayLowerBound", "50"),     
    ("EnableRadarTracks", "0"),      
    ("EnableAutoEngage", "0"),      
    ("ApplyDynamicTFollow", "100"), 
    ("ApplyDynamicTFollowApart", "100"), 
    ("ApplyDynamicTFollowDecel", "100"), 
    ("SccConnectedBus2", "0"),   
    ("TFollowRatio", "110"),
    ("JerkUpperLowerLimit", "8"),    
    ("KeepEngage", "1"),
    ("UseLaneLineSpeed", "0"),    
    ("PathOffset", "0"),  
    ("PathCostApply", "100"),
    ("HapticFeedbackWhenSpeedCamera", "0"),       
    ("MaxAngleFrames", "89"),       
    ("SoftHoldMode", "1"),       
    ("ApplyModelDistOrder", "32"),       
    ("TrafficStopAdjustRatio", "90"),       
    ("SteeringRateCost", "700"),       
    ("LateralMotionCost", "11"),       
    ("LateralAccelCost", "0"),       
    ("LateralJerkCost", "5"),       
    ("LateralTorqueKp", "100"),       
    ("LateralTorqueKi", "10"),       
    ("LateralTorqueKd", "0"),       
    ("LateralTorqueKf", "100"),       
    ("LateralTorqueCustom", "0"),       
    ("LateralTorqueAccelFactor", "2500"),       
    ("LateralTorqueFriction", "100"),       
    ("SteerActuatorDelay", "30"),       
    ("CruiseControlMode", "4"),
    ("CruiseOnDist", "0"),
    ("SteerRatioApply", "0"),
    ("SteerDeltaUp", "3"),       
    ("SteerDeltaDown", "7"),       
  ]
  if not PC:
    default_params.append(("LastUpdateTime", datetime.datetime.utcnow().isoformat().encode('utf8')))

  if params.get_bool("RecordFrontLock"):
    params.put_bool("RecordFront", True)

  # set unset params
  for k, v in default_params:
    if params.get(k) is None:
      params.put(k, v)

  # is this dashcam?
  if os.getenv("PASSIVE") is not None:
    params.put_bool("Passive", bool(int(os.getenv("PASSIVE", "0"))))

  if params.get("Passive") is None:
    raise Exception("Passive must be set to continue")

  # Create folders needed for msgq
  try:
    os.mkdir("/dev/shm")
  except FileExistsError:
    pass
  except PermissionError:
    print("WARNING: failed to make /dev/shm")

  # set version params
  params.put("Version", get_version())
  params.put("TermsVersion", terms_version)
  params.put("TrainingVersion", training_version)
  params.put("GitCommit", get_commit(default=""))
  params.put("GitBranch", get_short_branch(default=""))
  params.put("GitRemote", get_origin(default=""))

  # set dongle id
  reg_res = register(show_spinner=True)
  if reg_res:
    dongle_id = reg_res
  else:
    serial = params.get("HardwareSerial")
    raise Exception(f"Registration failed for device {serial}")
  os.environ['DONGLE_ID'] = dongle_id  # Needed for swaglog

  if not is_dirty():
    os.environ['CLEAN'] = '1'

  # init logging
  sentry.init(sentry.SentryProject.SELFDRIVE)
  cloudlog.bind_global(dongle_id=dongle_id, version=get_version(), dirty=is_dirty(),
                       device=HARDWARE.get_device_type())
  if os.path.isfile('/data/tmux_error.log'):
    os.remove('/data/tmux_error.log')


def manager_prepare() -> None:
  for p in managed_processes.values():
    p.prepare()


def manager_cleanup() -> None:
  # send signals to kill all procs
  for p in managed_processes.values():
    p.stop(block=False)

  # ensure all are killed
  for p in managed_processes.values():
    p.stop(block=True)

  cloudlog.info("everything is dead")


def manager_thread() -> None:

  if EON:
    Process(name="autoshutdownd", target=launcher, args=("selfdrive.autoshutdownd", "autoshutdownd")).start()
    system("am startservice com.neokii.optool/.MainService")

  Process(name="road_speed_limiter", target=launcher, args=("selfdrive.road_speed_limiter", "road_speed_limiter")).start()
  cloudlog.bind(daemon="manager")
  cloudlog.info("manager start")
  cloudlog.info({"environ": os.environ})

  params = Params()

  ignore: List[str] = []
  if params.get("DongleId", encoding='utf8') in (None, UNREGISTERED_DONGLE_ID):
    ignore += ["manage_athenad", "uploader"]
  if os.getenv("NOBOARD") is not None:
    ignore.append("pandad")
  ignore += [x for x in os.getenv("BLOCK", "").split(",") if len(x) > 0]

  ensure_running(managed_processes.values(), started=False, not_run=ignore)

  started_prev = False
  sm = messaging.SubMaster(['deviceState'])
  pm = messaging.PubMaster(['managerState'])

  print_timer = 0
  while True:
    sm.update()
    not_run = ignore[:]

    started = sm['deviceState'].started
    driverview = params.get_bool("IsDriverViewEnabled")
    ensure_running(managed_processes.values(), started, driverview, not_run)

    # trigger an update after going offroad
    if started_prev and not started and 'updated' in managed_processes:
      os.sync()
      managed_processes['updated'].signal(signal.SIGHUP)

    started_prev = started

    running = ' '.join("%s%s\u001b[0m" % ("\u001b[32m" if p.proc.is_alive() else "\u001b[31m", p.name)
                       for p in managed_processes.values() if p.proc)
    print_timer = (print_timer + 1)%10
    if print_timer == 0:
      print(running)
    cloudlog.debug(running)

    # send managerState
    msg = messaging.new_message('managerState')
    msg.managerState.processes = [p.get_process_state_msg() for p in managed_processes.values()]
    pm.send('managerState', msg)

    # Exit main loop when uninstall/shutdown/reboot is needed
    shutdown = False
    for param in ("DoUninstall", "DoShutdown", "DoReboot"):
      if params.get_bool(param):
        shutdown = True
        params.put("LastManagerExitReason", param)
        cloudlog.warning(f"Shutting down manager - {param} set")

    if shutdown:
      break


def main() -> None:
  preBuiltOn = Params().get_bool("OpkrPrebuiltOn")
  preBuiltFile = '/data/openpilot/prebuilt'
  if not os.path.isdir("/data/openpilot"):
      pass
  elif not os.path.isfile(preBuiltFile) and preBuiltOn:
    os.system("cd /data/openpilot; touch prebuilt")
  elif os.path.isfile(preBuiltFile) and not preBuiltOn:
    os.system("cd /data/openpilot; rm -f prebuilt")

  prepare_only = os.getenv("PREPAREONLY") is not None

  manager_init()
  os.system("python /data/openpilot/selfdrive/car/hyundai/values.py > /data/params/d/SupportedCars")
  os.system("python /data/openpilot/selfdrive/car/gm/values.py > /data/params/d/SupportedCars_gm")

  # Start UI early so prepare can happen in the background
  if not prepare_only:
    managed_processes['ui'].start()

  manager_prepare()

  if prepare_only:
    return

  # SystemExit on sigterm
  signal.signal(signal.SIGTERM, lambda signum, frame: sys.exit(1))

  try:
    manager_thread()
  except Exception:
    traceback.print_exc()
    sentry.capture_exception()
  finally:
    manager_cleanup()

  params = Params()
  if params.get_bool("DoUninstall"):
    cloudlog.warning("uninstalling")
    HARDWARE.uninstall()
  elif params.get_bool("DoReboot"):
    cloudlog.warning("reboot")
    HARDWARE.reboot()
  elif params.get_bool("DoShutdown"):
    cloudlog.warning("shutdown")
    HARDWARE.shutdown()


if __name__ == "__main__":
  unblock_stdout()

  try:
    main()
  except Exception:
    add_file_handler(cloudlog)
    cloudlog.exception("Manager failed to start")

    try:
      managed_processes['ui'].stop()
    except Exception:
      pass

    # Show last 3 lines of traceback
    error = traceback.format_exc(-3)
    error = "Manager failed to start\n\n" + error
    with TextWindow(error) as t:
      t.wait_for_exit()

    raise

  # manual exit because we are forked
  sys.exit(0)
